Correct PHP JSON_BIGINT_AS_STRING parameter and decode as an object

jackwakefield opened the pull request:
This fixes the placement of the `JSON_BIGINT_AS_STRING` option passed to `json_decode`, this is being passed as the `$assoc` parameter which decodes the JSON as an associative array.

I've changed the JSON to be decoded as as object rather as an array. This fixes an issue where the keys to a Protobuf map could be a numerical string, which then become integers when decoded as an associative array causing `expected string`

```php
$data = '{"1001": true, "foo": true}';

var_dump(json_decode($data, true));
// array(2) {
//   [1001]=>
//   bool(true)
//   ["foo"]=>
//   bool(true)
// }

var_dump(json_decode($data));
// object(stdClass)#1 (2) {
//   ["1001"]=>
//   bool(true)
//   ["foo"]=>
//   bool(true)
// }
```

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

bazel-io commented:
Can one of the admins verify this patch?

jackwakefield commented:
I signed it!

TeBoring commented:
ok to test

TeBoring commented:
Could you add a test to prevent regression?

jackwakefield commented:
I've added a simple test which will throw an exception when the JSON is being decoded as an associative array.

I've also changed the test script to run the tests against both the extension and PHP implementation as this was only reproducible with the latter. The PHP implementation tests are currently failing in `well_known_test.php` with the following, which I'll open an issue for.

```
Use of undefined constant url_prifix_len - assumed 'url_prifix_len'

php/src/Google/Protobuf/Any.php:212
php/tests/well_known_test.php:46
```

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

bazel-io commented:
Can one of the admins verify this patch?

TeBoring commented:
Could you rebase this change?

jackwakefield commented:
Done

TeBoring commented on php/src/Google/Protobuf/Internal/Message.php:
```diff
@@ -850,9 +850,9 @@ private function convertJsonValueToProtoValue(
     }
 
-    private function mergeFromJsonArray($array)
+    private function mergeFromJsonObject($object)
```
Is this rename intentional?

TeBoring replied:
The original mergeFromJsonArray is for implementation only. What's the benefit to use object instead of array?

jackwakefield replied:
When the field is a map PHPs type system will convert a key with string `"1"` to the number `1` when decoding as an associative array, which will cause an exception with Protobuf's type checking. Decoding as an object preserving the key types.

TeBoring commented on php/tests/test.sh:
```diff
@@ -12,11 +12,17 @@ tests=( array_test.php encode_decode_test.php generated_class_test.php generated
   php -dextension=../ext/google/protobuf/modules/protobuf.so `which phpunit` --bootstrap autoload.php $t
   echo ""
+
+  echo "************************************"
```
This is duplicate of what we do to test php implementation.

jackwakefield replied:
Could you point me in the right direction to run the existing tests? I wasn't able to get the tests to run against the PHP implementation (and the test to fail) without adding this.

TeBoring replied:
cd php
composer install
./vendor/bin/phpunit
